src/sync/gll.ts

```diff
@@ -0,0 +1,50 @@
+import PQueue from 'queue-promise';
+import { ILogger } from '../types';
+import { GlobalLoadLockTaskError } from '../errors';
+
+const MAX_CONCURRENT_LOAD_TASKS = 5;
+
+const GLL = new PQueue({ concurrent: MAX_CONCURRENT_LOAD_TASKS });
+
+/**
+ * Add task to the GLL and return a promise that will resolve or reject when the task resolves or rejects.
+ * @param task The task to execute.
+ * @param logger Logger instance to use.
+ */
+export function addTask(task: () => Promise<void>, logger: ILogger) {
+  const taskId = Math.random().toString(36).substring(2, 15);
+  const startTime = Date.now();
+  // This promise will resolve or reject when the task does, so the caller can know his task has ended.
+  const promise = new Promise<void>((resolve, reject) => {
+    GLL.on('resolve', data => {
+      if (data === taskId) {
+        resolve();
+      }
+    });
+
+    GLL.on('reject', data => {
+      if (data.taskId === taskId) {
+        reject(data.innerError);
+      }
+    });
+  });
+
+  /**
+   * Add the task to the queue and return or rejects with the taskId
+   * So the promise above can detect any issues and reject correctly.
+   */
+  GLL.enqueue(async () => {
+    try {
+      logger.info(`Task waited on queue for ${(Date.now() - startTime) / 1000}s`);
```

Comment:
Can you improve the log to add information about which wallet was waiting for that long? It might be useful for debugging.

Reply:
The wallet-ilb does not identify each wallet, they are just instances.
The idea behing the logger is that the wallet implementation can add a logger instance with more data.

For instance the wallet-headless will add the date and wallet-id as prefixes to the actual message.

Reply:
Got it. I'm wondering where we can store the queue time information so the client can log it properly. Should the wallet-lib provide an object with sync information (e.g., queue time, type, downloading time, processing time, and total time)?

Reply:
The wallet instance goes through the states:
- `CONNECTING`: Trying to establish a websocket connection with the fullnode.
- `SYNCING`: queue time + time to download the history.
- `PROCESSING`: Go through the history and process the balances and other metadata.
- `READY`: Finished processing

We could just keep an object that holds the timestamp when we last entered each state.
So knowning the queue time from this PR and the time between `SYNCING` and `PROCESSING` we can calculate how long it took to download the history.
From the time between `PROCESSING` and `READY` we know how long it took to process the history.
Since this is reacting to the state changes of the wallet we could just do this on the headless, what do you think?

Alternatively we could create a stats manager for the wallet like the stream stats manager from #750